### PR TITLE
alt proposal

### PR DESCRIPTION
**Pull request by dan9186**
- feat: batch actions for messages
- uuid validation for message ids
- buf validate
- add missing get path
- move get message action metadata to only be part of the get response
- move singular and batch actions to be represented as one endpoint

### CODE REVIEW

File: proto/o5/dempe/v1/service.proto [hunk 1]
```diff
@@ -18,9 +18,18 @@ service DempeService {
     };
   }
 
-  rpc GetMessage(GetMessageRequest) returns (GetMessageResponse) {}
+  rpc GetMessage(GetMessageRequest) returns (GetMessageResponse) {
+    option (google.api.http) = {
+      get: "/dempe/v1/message/{message_id}"
+    };
+  }
```
Comment (dan9186) [inline proto/o5/dempe/v1/service.proto:25]: Realized this was missing while working it over. If needed, we can pull this to a different PR.

File: proto/o5/dempe/v1/service.proto [hunk 2]
```diff
@@ -44,37 +53,41 @@ message GetMessageRequest {
...
+    google.protobuf.Timestamp timestamp = 2;
+    o5.auth.v1.Actor actor = 3;
+  }
+  repeated GetMessageAction actions = 2;
 }
 
 message MessageAction {
+  string id = 1 [
+    (buf.validate.field).string.uuid = true
+  ];
```
Comment (dan9186) [inline proto/o5/dempe/v1/service.proto:72]: I realized, this would actually be the action's ID, not the message ID. So I'm not sure how necessary this one is. Possibly should be removed.

File: proto/o5/dempe/v1/service.proto [hunk 3]
```diff
@@ -44,37 +53,41 @@ message GetMessageRequest {
...
 
   string note = 2;
 
-  o5.auth.v1.Actor actor = 3;
-
-  google.protobuf.Timestamp timestamp = 4;
```
Comment (dan9186) [inline proto/o5/dempe/v1/service.proto:61]: These have been moved to only exist as part of the `GetMessageResponse` as metadata about a singular action.

File: proto/o5/dempe/v1/service.proto [hunk 4]
```diff
@@ -44,37 +53,41 @@ message GetMessageRequest {
...
-
   oneof action {
     // optional
     ActionDelete delete = 10;
     ActionRequeue requeue = 11;
-    ActionEdit edit = 12;
+    ActionEdit edit = 12; // Will return an error if more than one id is provided
```
Comment (dan9186) [inline proto/o5/dempe/v1/service.proto:80]: If a singular message action request is to accept multiple ids of messages to act upon, this is a desirable software side validation. Not particularly a fan of it, but it is an alternative to having two endpoints for performing actions.

File: proto/o5/dempe/v1/service.proto (hunk 2 above)
Comment (joshuaslate) [reply proto/o5/dempe/v1/service.proto:72]: I was thinking it would make sense for the action to have an id, because it could appear in an "audit log." I assume there's a use case for showing actions on a message, because we're storing actor data for the actions. To that note, I'd imagine we should add another endpoint for a list of actions for a message.

Comment (daemonl) [reply proto/o5/dempe/v1/service.proto:72]: I haven't even considered storing an audit log. Seems like a good idea!

**Comment by dan9186** [comment]
Ok, I tried going through a few other considerations and iterations of things. This still cleans up things the best. As long as the singular endpoint is ok for 1 to N, then I'd say this is the lesser of the evils.

**Review by joshuaslate — APPROVED** [review]
